File: backend/src/routers/todos.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from src.database import get_db
from src.models import TodoItem
from src.schemas.todos import TodoItemResponse, CompleteTodoRequest, TodoItemUpdate
from src.auth import get_current_user

router = APIRouter(prefix="/todos", tags=["todos"])

NOT_FOUND_ERROR_CODE = 404
NOT_ACCESS_ERROR_CODE = 403


def _check_access(current_user: dict, item: TodoItem) -> bool:
    todolist = item.todolist
    if current_user["id"] == todolist.owner_id:
        return True
    if current_user["id"] in [member.id for member in todolist.members]:
        return True
    return False


@router.get("/{todo_id}", response_model=TodoItemResponse)
async def get_todo(
    todo_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> TodoItem:
    item = db.query(TodoItem).filter(TodoItem.id == todo_id).first()
    if not item:
        raise HTTPException(status_code=NOT_FOUND_ERROR_CODE, detail="Todo not found")
    if not _check_access(current_user, item):
        raise HTTPException(status_code=NOT_ACCESS_ERROR_CODE, detail="No access")
    return item


@router.delete("/{todo_id}")
async def delete_todo(
    todo_id: int,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> dict:
    item = db.query(TodoItem).filter(TodoItem.id == todo_id).first()
    if not item:
        raise HTTPException(status_code=NOT_FOUND_ERROR_CODE, detail="Todo not found")
    if not _check_access(current_user, item):
        raise HTTPException(status_code=NOT_ACCESS_ERROR_CODE, detail="No access")
    db.delete(item)
    db.commit()
    return {"deleted_id": item.id}


@router.patch("/{todo_id}", response_model=TodoItemResponse)
async def update_todo(
    todo_id: int,
    update_data: TodoItemUpdate,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> TodoItem:
    item = db.query(TodoItem).filter(TodoItem.id == todo_id).first()
    if not item:
        raise HTTPException(status_code=NOT_FOUND_ERROR_CODE, detail="Todo not found")
    if not _check_access(current_user, item):
        raise HTTPException(status_code=NOT_ACCESS_ERROR_CODE, detail="No access")

    updated_values = update_data.model_dump(exclude_unset=True)
    for key, value in updated_values.items():
        setattr(item, key, value)

    db.commit()
    return item


@router.patch("/{todo_id}/complete")
async def complete_todo(
    todo_id: int,
    complete_data: CompleteTodoRequest,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user),
) -> dict:
    item = db.query(TodoItem).filter(TodoItem.id == todo_id).first()
    if not item:
        raise HTTPException(status_code=NOT_FOUND_ERROR_CODE, detail="Todo not found")
    if not _check_access(current_user, item):
        raise HTTPException(status_code=NOT_ACCESS_ERROR_CODE, detail="No access")

    item.is_completed = complete_data.is_completed
    db.commit()
    return {"completed_id": item.id}

File: frontend/src/api/todos.ts
import type {
	HandleTodoCompleteRequest,
	Todo,
	TodoCreatePayload,
	TodoUpdatePayload,
} from "../types";
import client from "./client";

export const getTodos = (wishlistId: number): Promise<Todo[]> =>
	client.get(`/wishlists/${wishlistId}/todos`).then((r) => r.data);

export const getTodo = (todoId: number): Promise<Todo> =>
	client.get(`/todos/${todoId}`).then((r) => r.data);

export const createTodo = (data: TodoCreatePayload): Promise<Todo> =>
	client.post(`/wishlists/${data.todolist_id}/todos`, data).then((r) => r.data);

export const updateTodo = (
	todoId: number,
	data: TodoUpdatePayload,
): Promise<Todo> => client.patch(`/todos/${todoId}`, data).then((r) => r.data);

export const handleTodoComplete = (
	todoId: number,
	data: HandleTodoCompleteRequest,
): Promise<{ completed_id: number }> =>
	client.patch(`/todos/${todoId}/complete`, data).then((r) => r.data);

export const deleteTodo = (todoId: number): Promise<{ deleted_id: number }> =>
	client.delete(`/todos/${todoId}`).then((r) => r.data);
